Add tips about service client and rate (#4132)
ros2/ros2_documentation: source/How-To-Guides/Migrating-from-ROS1/Migrating-Python-Packages.rst

       node.get_logger().info('service not available, waiting again...')
   resp = add_two_ints.call_async(req)
   rclpy.spin_until_future_complete(node, resp)

.. warning::

>   Synchronous calls such as ``rclpy.spin_until_future_complete`` can cause deadlock.
>   For more details see `Sync deadlock <https://docs.ros.org/en/{DISTRO}/How-To-Guides/Sync-Vs-Async.html#sync-deadlock>`__.

Reviewer: ```suggestion
   Synchronous calls such as ``rclpy.spin_until_future_complete`` can cause deadlock.
   For more details see the :doc:`sync deadlock article <../Sync-Vs-Async>`.
```

       node.get_logger().info('service not available, waiting again...')
   resp = add_two_ints.call_async(req)
   rclpy.spin_until_future_complete(node, resp)

Be careful not to run `rclpy.spin_until_future_complete` from within a ROS callback function.
For more details please see `Sync deadlock <https://docs.ros.org/en/{DISTRO}/How-To-Guides/Sync-Vs-Async.html#sync-deadlock>`__.

Executing at a Specific Rate
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In ROS 1:

.. code-block:: python

   rate = rospy.Rate(2)
   while not rospy.is_shutdown():
      msg.data += math_stuff()
      pub.publish(msg)
      rate.sleep()

In ROS 2:

.. code-block:: python

   node.create_rate(2)
   while rclpy.ok():
      msg.data += math_stuff()
      node.pub.publish(msg)

>       rclpy.spin_once(node)

Reviewer 2: @mjforan this `Rate` example is not gonna work as expected if any other executables are available in the executor... for example, if the node is subscribing some higher frequency topic than this rate (this is likely), it will not sleep the period 2 Hz. that can be useful in some cases as described below, but i am now skeptical if we want to introduce this `Rate` here?

probably we want to guide the user to use `create_timer` instead, and publish the data in the timer callback? that would be the better guidance for the migration here?

CC: @clalancette

Author: Good catch. However, running things in the timer callback does not work either for things like a service client call. So we will have to make a timer callback that sets a flag. That way the worker function can tell if the wakeup was triggered by the timer or a higher frequency callback. I'm away from my computer this weekend but I'll write up a revised example next week.

Reviewer 2: > However, running things in the timer callback does not work either for things like a service client call

if that is sync call request. maybe that is the best practice or recommended guide for this migration doc.
the reason i am concerned with this is, that many user will read this doc especially for the migration phase, they would use the introduced sample almost as it is for the migration. we of course want them to avoid the possible problems...

~~btw, i really do not understand how to use this `Rate` class... that only makes sense if we have dedicated node with `Rate` and `spin`... i do not find useful this `Rate` class, but that is not the point here.~~

sorry for taking time on this, really appreciate your effort.

@clalancette what do you think? any thoughts?

Author: @fujitatomoya I updated the rate/timer example and updated the explanation text.

       node.get_logger().info('service not available, waiting again...')
   resp = add_two_ints.call_async(req)
   rclpy.spin_until_future_complete(node, resp)

Be careful not to run `rclpy.spin_until_future_complete` from within a ROS callback function! For more details `see here. <https://docs.ros.org/en/rolling/How-To-Guides/Sync-Vs-Async.html#sync-deadlock/>`_

> 

Reviewer 2: ```suggestion
Be careful not to run `rclpy.spin_until_future_complete` from within a ROS callback function.
For more details please see `Sync deadlock <https://docs.ros.org/en/{DISTRO}/How-To-Guides/Sync-Vs-Async.html#sync-deadlock>`__.
```

Executing at a Specific Rate
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In ROS 1:

.. code-block:: python

   rate = rospy.Rate(2)
   while not rospy.is_shutdown():
      msg.data += math_stuff()
      pub.publish(msg)
      rate.sleep()

In ROS 2:

.. code-block:: python

   node.create_rate(2)
   while rclpy.ok():
      msg.data += math_stuff()
      node.pub.publish(msg)

>       rclpy.spin_once(node)

Reviewer 2: Does this really work? I think this is supposed to be like this.

```suggestion
   rate = node.create_rate(2)
   while rclpy.ok():
      msg.data += math_stuff()
      node.pub.publish(msg)
      rate.sleep()
```

Author: Yes, that is how it works in ROS 2. It is indeed counterintuitive, which is why I included it here. `rate.sleep()` relies on the executor spinning so if you `sleep` manually like that it will never wake up. More info here:
https://answers.ros.org/question/358343/rate-and-sleep-function-in-rclpy-library-for-ros2/

Reviewer 2: Ah, i see. i was confused with `rclcpp` implementation. thanks for the explanation.
as you mentioned, `rate.sleep()` requires `executor` is spinning, i think this is tricky but that is the current implementation.
that said, i am also confused 😅 so that it is probably add this sample in here 👍 

How about holding the `rate` object in stack? so that user can destroy that later? it would be better practice to introduce here? (NOTE: according to the current code, it does not destroy `Timer` underneath, so i do not think that works at this moment. but this is another issue, i will create the issue for this later.)

```suggestion
   rate = node.create_rate(2)
   while rclpy.ok():
      msg.data += math_stuff()
      node.pub.publish(msg)
      rclpy.spin_once(node)
```

Author: That belongs in a separate example of how to start, stop, and destroy timers; it does not add anything here. The timer will be destroyed when the process ends and that is fine for this simple node.

Reviewer 2: i was not saying that we introduce destroy here, but holding the rate object. i do not have strong opinion for this. just thought it would be nicer for the reference code since it is likely that people copy&paste the example. anyway, this is really minor comment, resolving.

If ``create_rate`` is omitted, the while loop will sleep after entering ``spin_once`` and will continue after a callback is processed.

> This could be useful if you have a worker function later in the loop which is waiting for some data from ROS.

Reviewer 2: can you elaborate this? i am not really following this? with `create_rate`, `spin_once` will sleep for 2 seconds?

Author: "omit" means "leave out". So _without_ `create_rate`,  `spin_once` will wait indefinitely for a callback to process. Could be useful if you have a node that waits for data to process, does some work, then goes back to sleep. It doesn't need to wake up at a regular interval, so you can use this pattern to sleep and automatically wake up when there is work to be done.

Author: How about this:

If `create_rate` is removed, `spin_once` will sleep indefinitely until a callback is activated by incoming ROS data. This could be useful for a node which: sleeps until data is received, does some work, then goes back to sleep. Instead of waking at a regular interval and checking for data it can automatically wake up whenever data is received.

       node.get_logger().info('service not available, waiting again...')
   resp = add_two_ints.call_async(req)
   rclpy.spin_until_future_complete(node, resp)

**Important:** ``rclpy.spin_until_future_complete`` cannot be executed under the context of another ``rclpy.spin``.
This means the function cannot be used within e.g. a ROS 2 subscriber callback. To avoid this issue you can store

> data from the callback and run the service call later (outside of ``spin``).

Reviewer: Does this work in ROS 1?  I think we should only put this here if this used to work, and doesn't anymore.

Author: Yes, I just verified using Noetic.

       node.get_logger().info('service not available, waiting again...')
   resp = add_two_ints.call_async(req)
   rclpy.spin_until_future_complete(node, resp)

Be careful not to run ``rclpy.spin_until_future_complete`` from within a ROS callback function.
For more details please see `Sync deadlock <https://docs.ros.org/en/{DISTRO}/How-To-Guides/Sync-Vs-Async.html#sync-deadlock>`__.

Executing at a Specific Rate
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In ROS 1:

.. code-block:: python

   rate = rospy.Rate(2)
   while not rospy.is_shutdown():
      msg.data += math_stuff()
      pub.publish(msg)
      rate.sleep()

In ROS 2:

.. code-block:: python

   node.create_timer(1/2, math_stuff)
   rclpy.spin()

> What if the helper function ``math_stuff`` needs to call a service (cannot be done in a callback)?

Reviewer 2: I think we can go more straight forward to map the publisher with timer example directly against ROS 1 above? `Sync deadlock` note is already added in service section, so i do not think we duplicate it again here.

i am not inclined to introduce `flag` here, it does not seem relevant this example. besides, i think we should use https://github.com/ros2/rclpy/blob/7d52cfa9801597c62be98f80d1a2c769f9768fff/rclpy/rclpy/timer.py#L97-L99 instead?
